lib/src/rules/invariant_boolean.dart

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.invariant_boolean;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/linter.dart';
+import 'package:linter/src/util/boolean_expression_utilities.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+import 'package:linter/src/util/tested_expressions.dart';
+
+const _desc = r'Conditions should not unconditionally evaluate to "TRUE" or to "FALSE"';
+
+const _details = r'''
+
+**DON'T** test for conditions that can be inferred at compile time or test the
+same condition twice.
+Conditional statements using a condition which cannot be anything but FALSE have
+the effect of making blocks of code non-functional. If the condition cannot
+evaluate to anything but TRUE, the conditional statement is completely
+redundant, and makes the code less readable.
+It is quite likely that the code does not match the programmer's intent.
+Either the condition should be removed or it should be updated so that it does
+not always evaluate to TRUE or FALSE and does not perform redundant tests.
+This rule will hint to the test conflicting with the linted one.
+
+**BAD:**
+```
+//foo can't be both equal and not equal to bar in the same expression
+if(foo == bar && something && foo != bar) {...}
+```
+
+**BAD:**
+```
+void compute(int foo) {
+  if (foo == 4) {
+    doSomething();
+    // We know foo is equal to 4 at this point, so the next condition is always false
+    if (foo > 4) {...}
+    ...
+  }
+  ...
+}
+```
+
+**BAD:**
+```
+void compute(boolean foo) {
+  if (foo) {
+    return;
+  }
+  doSomething();
+  // foo is always false here
+  if (foo){...}
+  ...
+}
+```
+
+**GOOD:**
+```
+void nestedOK() {
+  if (foo == bar) {
+    foo = baz;
+    if (foo != bar) {...}
+  }
+}
+```
+
+**GOOD:**
+```
+void nestedOk2() {
+  if (foo == bar) {
+    return;
+  }
+
+  foo = baz;
+  if (foo == bar) {...} // OK
+}
+```
+
+**GOOD:**
+```
+void nestedOk5() {
+  if (foo != null) {
+    if (bar != null) {
+      return;
+    }
+  }
+
+  if (bar != null) {...} // OK
+}
+```
+''';
+
+
+class InvariantBoolean extends LintRule {
+  _Visitor _visitor;
+
+  InvariantBoolean() : super(
+      name: 'invariant_boolean',
+      description: _desc,
+      details: _details,
+      group: Group.errors,
+      maturity: Maturity.experimental) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  visitBinaryExpression(BinaryExpression node) {
```

Comment:
It would be more efficient to only visit binary expressions when they are children of an `if` statement than to test the parent of every binary expression. (That is, don't use a visit method for binary expressions, but invoke a helper method, or maybe just inline the code, when the condition is a binary expression.)